File: frontend/todo-app/src/models/todo-item.ts
/**
 * Represents a single todo item.
 * @module models/todo-item
 */
export interface TodoItem {
  /** Unique identifier for the todo item. */
  id: string;
  /** Title or description of the todo item. */
  title: string;
  /** Whether the todo item is completed. */
  isCompleted: boolean;
  /** Date and time when the todo item was created. */
  createdDate: string;
  /**
   * When a todo is completed the API returns the timestamp at which it
   * was completed.  This value is optional because uncompleted items
   * don't have a completion date.  Consumers should treat `null` as
   * “not completed yet”.
   */
  completedDate?: string | null;
  /**
   * Optional calendar date for when the task is due.  Stored as an ISO
   * 8601 date string (YYYY‑MM‑DD) or `null` if no date has been
   * assigned.  A date in the past will be rejected by the API.
   */
  dueDate?: string | null;
  /**
   * Optional time on the due date when the task should be completed.
   * Stored as an ISO 8601 time string (HH:mm or HH:mm:ss) or `null` if
   * no time has been assigned.  If present, a `dueDate` should also be
   * provided.
   */
  dueTime?: string | null;
  /** Identifier for the creator of the todo item. */
  createdBy: string;
}

File: backend/Todo.Api/Models/Domain/TodoItem.cs
using Todo.Api.Models.Contracts;

namespace Todo.Api.Models.Domain;

/// <summary>
/// Domain model representing a Todo item.
/// </summary>
public record TodoItem
{
    /// <summary>
    /// Maximum length for the title of a Todo item.
    /// </summary>
    public const int MaxTitleLength = 240;
    /// <summary>
    /// The unique identifier of the Todo item.
    /// </summary>
    public Guid Id { get; set; }
    /// <summary>
    /// The title of the Todo item.
    /// </summary>
    public string Title { get; set; } = string.Empty;
    /// <summary>
    /// Indicates whether the Todo item is completed.
    /// </summary>
    public bool IsCompleted { get; set; }
    /// <summary>
    /// The date and time when the Todo item was created.
    /// </summary>
    public DateTimeOffset CreatedDate { get; set; }
    /// <summary>
    /// The date and time when the Todo item was completed.
    /// </summary>
    public DateTimeOffset? CompletedDate { get; set; }
    /// <summary>
    /// The date on which the Todo item is due. Date-only to allow setting
    /// a calendar date without requiring a time of day.
    /// </summary>
    public DateOnly? DueDate { get; set; }
    /// <summary>
    /// The time of day at which the Todo item is due. Time-only to allow
    /// fine‑grained control over deadlines without affecting the calendar date.
    /// </summary>
    public TimeOnly? DueTime { get; set; }
    /// <summary>
    /// The identifier of the user who created the Todo item.
    /// </summary>
    public Guid CreatedBy { get; set; }

    public TodoItemResponseDTO ToDTO()
    {
        return new TodoItemResponseDTO
        {
            Id = Id,
            Title = Title,
            IsCompleted = IsCompleted,
            CreatedDate = CreatedDate,
            CompletedDate = CompletedDate,
            DueDate = DueDate,
            DueTime = DueTime,
            CreatedBy = CreatedBy
        };
    }
}